File: signaling.js
const WebSocket = require("ws");

const webSocketPort = 8080;
const server = new WebSocket.Server({ port: webSocketPort });

// roomId -> { offers: Map<peerId, socket>, answers: Map<peerId, socket> }
const rooms = new Map();
const LEGACY_ROOM_ID = "__legacy_single_pair__";

const getRoom = (roomId) => {
  const id = roomId || LEGACY_ROOM_ID;
  if (!rooms.has(id)) {
    rooms.set(id, { offers: new Map(), answers: new Map() });
  }
  return rooms.get(id);
};

const safeSend = (socket, payload) => {
  if (!socket || socket.readyState !== WebSocket.OPEN) return false;
  socket.send(JSON.stringify(payload));
  return true;
};

const normalizeRole = (role) => (role === "offer" || role === "answer" ? role : null);

const attachMeta = (socket, meta) => {
  socket.__meta = meta;
};

const detachMeta = (socket) => {
  delete socket.__meta;
};

const registerClient = (socket, { role, roomId, peerId, legacy = false }) => {
  const normRole = normalizeRole(role);
  if (!normRole) {
    safeSend(socket, { type: "error", message: "Invalid role; expected offer|answer" });
    return;
  }
  const id = peerId || (legacy ? `legacy-${normRole}` : null);
  if (!id) {
    safeSend(socket, { type: "error", message: "peerId is required" });
    return;
  }

  const room = getRoom(roomId);
  const targetMap = normRole === "offer" ? room.offers : room.answers;

  // Replace existing connection for the same peerId (if any)
  const prev = targetMap.get(id);
  if (prev && prev !== socket) {
    try {
      prev.close();
    } catch (_) {}
  }

  targetMap.set(id, socket);
  attachMeta(socket, { role: normRole, roomId: roomId || LEGACY_ROOM_ID, peerId: id });

  safeSend(socket, { type: "registered", role: normRole, roomId: roomId || LEGACY_ROOM_ID, peerId: id });
  console.log(`[register] role=${normRole} room=${roomId || LEGACY_ROOM_ID} peer=${id}`);
};

const routeSignal = ({ from, to, roomId, data }) => {
  const room = getRoom(roomId);
  const targetMap = to.role === "offer" ? room.offers : room.answers;
  const targetSocket = targetMap.get(to.peerId);

  if (!targetSocket) {
    console.warn(`[signal] target not found room=${roomId} role=${to.role} peer=${to.peerId}`);
    return;
  }

  safeSend(targetSocket, {
    type: "signal",
    roomId,
    from,
    to,
    data,
  });
  console.log(`[signal] ${from.role}:${from.peerId} -> ${to.role}:${to.peerId} room=${roomId} kind=${data && data.kind}`);
};

const handleLegacyRelay = (socket, type, payload) => {
  // Legacy schema: single offer/answer pair, message type decides direction.
  const room = getRoom(LEGACY_ROOM_ID);
  const targetRole = type.includes("offer") ? "answer" : "offer";
  const targetMap = targetRole === "offer" ? room.offers : room.answers;
  const targetSocket = targetMap.values().next().value; // first (only) entry
  if (!targetSocket) {
    console.warn(`[legacy] target not connected role=${targetRole}`);
    return;
  }
  safeSend(targetSocket, { type, payload });
  console.log(`[legacy] relayed ${type} to ${targetRole}`);
};

server.on("connection", (socket) => {
  socket.on("message", (message) => {
    let data;
    try {
      data = JSON.parse(message);
    } catch (error) {
      safeSend(socket, { type: "error", message: "Invalid JSON" });
      return;
    }

    const { type, payload } = data;
    switch (type) {
      case "register": {
        const { role, roomId, peerId } = payload || {};
        registerClient(socket, { role, roomId, peerId });
        break;
      }
      case "signal": {
        // Accept both new-style envelope (top-level fields) and old payload-wrapped shape.
        const envelope = payload || data;
        const { roomId, from, to, data: signalData } = envelope || {};
        if (!from || !to || !from.role || !to.role || !to.peerId) {
          safeSend(socket, { type: "error", message: "signal requires from/to with role and to.peerId" });
          break;
        }
        if (!signalData) {
          safeSend(socket, { type: "error", message: "signal requires data" });
          break;
        }
        const fromRole = normalizeRole(from.role);
        const toRole = normalizeRole(to.role);
        if (!fromRole || !toRole) {
          safeSend(socket, { type: "error", message: "signal roles must be offer|answer" });
          break;
        }
        routeSignal({
          from: { role: fromRole, peerId: from.peerId },
          to: { role: toRole, peerId: to.peerId },
          roomId: roomId || LEGACY_ROOM_ID,
          data: signalData,
        });
        break;
      }
      // Legacy compatibility (single pair)
      case "register-offer":
        registerClient(socket, { role: "offer", roomId: LEGACY_ROOM_ID, peerId: "legacy-offer", legacy: true });
        break;
      case "register-answer":
        registerClient(socket, { role: "answer", roomId: LEGACY_ROOM_ID, peerId: "legacy-answer", legacy: true });
        break;
      case "offer":
      case "answer":
      case "ice-offer":
      case "ice-answer":
        handleLegacyRelay(socket, type, payload);
        break;
      default:
        safeSend(socket, { type: "error", message: `Unknown type: ${type}` });
        console.warn(`[warn] unknown message type: ${type}`);
        break;
    }
  });

  socket.on("close", () => {
    const meta = socket.__meta;
    if (meta && meta.roomId) {
      const room = rooms.get(meta.roomId);
      if (room) {
        const targetMap = meta.role === "offer" ? room.offers : room.answers;
        targetMap.delete(meta.peerId);
        if (room.offers.size === 0 && room.answers.size === 0) {
          rooms.delete(meta.roomId);
        }
      }
      console.log(`[disconnect] role=${meta.role} room=${meta.roomId} peer=${meta.peerId}`);
    }
    detachMeta(socket);
  });
});

console.log(`WebSocket server running on ws://localhost:${webSocketPort}`);

File: signaling.rs
use axum::{
    extract::ws::{Message, WebSocket, WebSocketUpgrade},
    response::IntoResponse,
    routing::get,
    Router,
};
use futures_util::{SinkExt, StreamExt};
use serde::{Deserialize, Serialize};
use serde_json::{json, Value};
use std::{
    collections::HashMap,
    net::SocketAddr,
    sync::{
        atomic::{AtomicU64, Ordering},
        Arc,
    },
};
use tokio::sync::{mpsc, oneshot, RwLock};

const WS_PORT: u16 = 8080;
const LEGACY_ROOM_ID: &str = "__legacy_single_pair__";

static CONN_ID: AtomicU64 = AtomicU64::new(1);

#[derive(Clone, Debug)]
struct PeerConn {
    conn_id: u64,
    tx: mpsc::UnboundedSender<Message>,
    close_tx: Arc<tokio::sync::Mutex<Option<oneshot::Sender<()>>>>,
}

#[derive(Default, Debug)]
struct Room {
    offers: HashMap<String, PeerConn>,
    answers: HashMap<String, PeerConn>,
}

#[derive(Default)]
struct AppState {
    rooms: RwLock<HashMap<String, Room>>,
}

#[derive(Clone, Debug)]
struct Meta {
    conn_id: u64,
    role: Role,
    room_id: String,
    peer_id: String,
}

#[derive(Clone, Debug, PartialEq, Eq)]
enum Role {
    Offer,
    Answer,
}

impl Role {
    fn normalize(s: &str) -> Option<Self> {
        match s {
            "offer" => Some(Role::Offer),
            "answer" => Some(Role::Answer),
            _ => None,
        }
    }
    fn as_str(&self) -> &'static str {
        match self {
            Role::Offer => "offer",
            Role::Answer => "answer",
        }
    }
}

#[derive(Debug, Deserialize)]
struct IncomingEnvelope {
    #[serde(rename = "type")]
    typ: String,
    payload: Option<Value>,
}

#[tokio::main]
async fn main() {
    let state = Arc::new(AppState::default());

    let app = Router::new()
        .route("/ws", get({
            let state = state.clone();
            move |ws: WebSocketUpgrade| {
                let state = state.clone();
                async move { ws.on_upgrade(move |socket| handle_socket(socket, state)) }
            }
        }));

    let addr = SocketAddr::from(([0, 0, 0, 0], WS_PORT));
    println!("WebSocket server running on ws://localhost:{}/ws", WS_PORT);

    let listener = tokio::net::TcpListener::bind(addr).await.unwrap();
    axum::serve(listener, app).await.unwrap();
}

async fn handle_socket(mut socket: WebSocket, state: Arc<AppState>) {
    let conn_id = CONN_ID.fetch_add(1, Ordering::Relaxed);

    // outgoing queue
    let (out_tx, mut out_rx) = mpsc::unbounded_channel::<Message>();

    // close signal (to force-close old connections on replace)
    let (close_tx, mut close_rx) = oneshot::channel::<()>();
    let close_tx = Arc::new(tokio::sync::Mutex::new(Some(close_tx)));

    // sender task
    let (mut ws_sender, mut ws_receiver) = socket.split();

    let sender_task = tokio::spawn(async move {
        while let Some(msg) = out_rx.recv().await {
            if ws_sender.send(msg).await.is_err() {
                break;
            }
        }
    });

    // per-connection meta
    let mut meta: Option<Meta> = None;

    // helper: send JSON to this socket
    let send_self = |v: Value| -> bool {
        out_tx.send(Message::Text(v.to_string())).is_ok()
    };

    // receive loop
    loop {
        tokio::select! {
            _ = &mut close_rx => {
                // forced close
                let _ = ws_receiver.close().await;
                break;
            }
            maybe = ws_receiver.next() => {
                let Some(Ok(msg)) = maybe else { break; };

                match msg {
                    Message::Text(text) => {
                        let parsed: Result<IncomingEnvelope, _> = serde_json::from_str(&text);
                        let Ok(env) = parsed else {
                            send_self(json!({"type":"error","message":"Invalid JSON"}));
                            continue;
                        };

                        match env.typ.as_str() {
                            "register" => {
                                let payload = env.payload.unwrap_or(Value::Null);
                                let role_s = payload.get("role").and_then(|v| v.as_str()).unwrap_or("");
                                let room_id = payload.get("roomId").and_then(|v| v.as_str()).unwrap_or(LEGACY_ROOM_ID);
                                let peer_id = payload.get("peerId").and_then(|v| v.as_str()).unwrap_or("");

                                register_client(
                                    conn_id,
                                    &state,
                                    &out_tx,
                                    &close_tx,
                                    &mut meta,
                                    role_s,
                                    room_id,
                                    peer_id,
                                    false,
                                ).await;
                            }
                            "signal" => {
                                let payload = env.payload.unwrap_or(Value::Null);
                                let room_id = payload.get("roomId").and_then(|v| v.as_str()).unwrap_or(LEGACY_ROOM_ID).to_string();

                                let from = payload.get("from").cloned().unwrap_or(Value::Null);
                                let to = payload.get("to").cloned().unwrap_or(Value::Null);
                                let data = payload.get("data").cloned();

                                let Some(data) = data else {
                                    send_self(json!({"type":"error","message":"signal requires data"}));
                                    continue;
                                };

                                let from_role_s = from.get("role").and_then(|v| v.as_str()).unwrap_or("");
                                let to_role_s = to.get("role").and_then(|v| v.as_str()).unwrap_or("");
                                let to_peer = to.get("peerId").and_then(|v| v.as_str()).unwrap_or("");
                                let from_peer = from.get("peerId").and_then(|v| v.as_str()).unwrap_or("");

                                let Some(from_role) = Role::normalize(from_role_s) else {
                                    send_self(json!({"type":"error","message":"signal roles must be offer|answer"}));
                                    continue;
                                };
                                let Some(to_role) = Role::normalize(to_role_s) else {
                                    send_self(json!({"type":"error","message":"signal roles must be offer|answer"}));
                                    continue;
                                };
                                if to_peer.is_empty() {
                                    send_self(json!({"type":"error","message":"signal requires from/to with role and to.peerId"}));
                                    continue;
                                }

                                route_signal(
                                    &state,
                                    room_id,
                                    from_role,
                                    from_peer.to_string(),
                                    to_role,
                                    to_peer.to_string(),
                                    data,
                                ).await;
                            }

                            // Legacy compatibility
                            "register-offer" => {
                                register_client(
                                    conn_id,
                                    &state,
                                    &out_tx,
                                    &close_tx,
                                    &mut meta,
                                    "offer",
                                    LEGACY_ROOM_ID,
                                    "legacy-offer",
                                    true,
                                ).await;
                            }
                            "register-answer" => {
                                register_client(
                                    conn_id,
                                    &state,
                                    &out_tx,
                                    &close_tx,
                                    &mut meta,
                                    "answer",
                                    LEGACY_ROOM_ID,
                                    "legacy-answer",
                                    true,
                                ).await;
                            }
                            "offer" | "answer" | "ice-offer" | "ice-answer" => {
                                let payload = env.payload.unwrap_or(Value::Null);
                                handle_legacy_relay(&state, env.typ, payload).await;
                            }

                            other => {
                                send_self(json!({"type":"error","message": format!("Unknown type: {}", other)}));
                                eprintln!("[warn] unknown message type: {}", other);
                            }
                        }
                    }
                    Message::Close(_) => break,
                    Message::Binary(_) => {
                        // ignore (signalingはText/JSON前提)
                    }
                    _ => {}
                }
            }
        }
    }

    // cleanup on disconnect
    if let Some(m) = meta.take() {
        cleanup_disconnect(&state, &m).await;
        println!(
            "[disconnect] role={} room={} peer={}",
            m.role.as_str(),
            m.room_id,
            m.peer_id
        );
    }

    // stop sender task
    sender_task.abort();
}

async fn register_client(
    conn_id: u64,
    state: &Arc<AppState>,
    out_tx: &mpsc::UnboundedSender<Message>,
    close_tx: &Arc<tokio::sync::Mutex<Option<oneshot::Sender<()>>>>,
    meta: &mut Option<Meta>,
    role_s: &str,
    room_id_in: &str,
    peer_id_in: &str,
    _legacy: bool,
) {
    let Some(role) = Role::normalize(role_s) else {
        let _ = out_tx.send(Message::Text(
            json!({"type":"error","message":"Invalid role; expected offer|answer"}).to_string(),
        ));
        return;
    };

    let room_id = if room_id_in.is_empty() { LEGACY_ROOM_ID } else { room_id_in }.to_string();
    let peer_id = peer_id_in.to_string();
    if peer_id.is_empty() {
        let _ = out_tx.send(Message::Text(
            json!({"type":"error","message":"peerId is required"}).to_string(),
        ));
        return;
    }

    // build current connection handle
    let conn = PeerConn {
        conn_id,
        tx: out_tx.clone(),
        close_tx: close_tx.clone(),
    };

    // insert & replace existing (force close old)
    {
        let mut rooms = state.rooms.write().await;
        let room = rooms.entry(room_id.clone()).or_insert_with(Room::default);
        let target_map = match role {
            Role::Offer => &mut room.offers,
            Role::Answer => &mut room.answers,
        };

        if let Some(prev) = target_map.insert(peer_id.clone(), conn.clone()) {
            if prev.conn_id != conn_id {
                // ask old connection to close (safe even if already closed)
                if let Some(tx) = prev.close_tx.lock().await.take() {
                    let _ = tx.send(());
                }
            }
        }
    }

    *meta = Some(Meta {
        conn_id,
        role: role.clone(),
        room_id: room_id.clone(),
        peer_id: peer_id.clone(),
    });

    let _ = out_tx.send(Message::Text(
        json!({
            "type":"registered",
            "role": role.as_str(),
            "roomId": room_id,
            "peerId": peer_id
        })
        .to_string(),
    ));

    println!(
        "[register] role={} room={} peer={}",
        role.as_str(),
        room_id,
        peer_id
    );
}

async fn route_signal(
    state: &Arc<AppState>,
    room_id: String,
    from_role: Role,
    from_peer: String,
    to_role: Role,
    to_peer: String,
    data: Value,
) {
    let target = {
        let rooms = state.rooms.read().await;
        let Some(room) = rooms.get(&room_id) else {
            eprintln!("[signal] room not found room={}", room_id);
            return;
        };
        let map = match to_role {
            Role::Offer => &room.offers,
            Role::Answer => &room.answers,
        };
        map.get(&to_peer).cloned()
    };

    let Some(target_conn) = target else {
        eprintln!(
            "[signal] target not found room={} role={} peer={}",
            room_id,
            to_role.as_str(),
            to_peer
        );
        return;
    };

    let payload = json!({
        "type": "signal",
        "roomId": room_id,
        "from": { "role": from_role.as_str(), "peerId": from_peer },
        "to":   { "role": to_role.as_str(),   "peerId": to_peer },
        "data": data
    });

    let _ = target_conn.tx.send(Message::Text(payload.to_string()));
}

async fn handle_legacy_relay(state: &Arc<AppState>, typ: String, payload: Value) {
    let target_role = if typ.contains("offer") { Role::Answer } else { Role::Offer };

    let target = {
        let rooms = state.rooms.read().await;
        let Some(room) = rooms.get(LEGACY_ROOM_ID) else {
            eprintln!("[legacy] room not found");
            return;
        };
        let map = match target_role {
            Role::Offer => &room.offers,
            Role::Answer => &room.answers,
        };
